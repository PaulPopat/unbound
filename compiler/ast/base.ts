import { Location } from "@compiler/location";

export abstract class Visitor<T extends Component> {
  abstract get OperatesOn(): Array<new (...args: any[]) => T>;

  abstract Visit(target: T): Component | undefined;
}

export abstract class Component {
  readonly #location: Location;

  constructor(location: Location) {
    this.#location = location;
  }

  get Location() {
    return this.#location;
  }

  abstract inner_visited(visitor: Visitor<Component>): Component;

  visited(visitor: Visitor<Component>): Component {
    if (!visitor.OperatesOn.find((o) => this instanceof o))
      return this.inner_visited(visitor);

    const result = visitor.Visit(this);
    if (!result) return this.inner_visited(visitor);

    return result;
  }

  type_safe_visited<T extends Component>(
    checker: abstract new (...args: any[]) => T,
    visitor: Visitor<Component>
  ) {
    const result = this.visited(visitor);
    if (!(result instanceof checker))
      throw new Error("Invalid type from visitor");

    return result;
  }

  abstract get type_name(): string;

  abstract get extra_json(): Record<never, never>;

  get json(): unknown {
    return {
      ...this.extra_json,
      type: this.type_name,
      location: this.#location.json,
    };
  }
}

export class ComponentGroup<TComponent extends Component> {
  readonly #components: Array<TComponent>;

  constructor(...components: Array<TComponent>) {
    this.#components = components;
  }

  get Length() {
    return this.#components.length;
  }

  get First() {
    return this.#components[0];
  }

  get Last() {
    return this.#components[this.#components.length - 1];
  }

  get Location() {
    return new Location(
      this.First.Location.FileName,
      this.First.Location.StartLine,
      this.First.Location.StartColumn,
      this.Last.Location.EndLine,
      this.Last.Location.EndColumn
    );
  }

  visited(visitor: Visitor<Component>) {
    return new ComponentGroup(
      ...this.#components.map((c) => c.visited(visitor))
    );
  }

  type_safe_visited<T extends Component>(
    checker: abstract new (...args: any[]) => T,
    visitor: Visitor<Component>
  ) {
    return new ComponentGroup(
      ...this.#components.map((c) => c.type_safe_visited(checker, visitor))
    );
  }

  get json() {
    return this.#components.map((c) => c.json);
  }

  iterator() {
    return this.#components[Symbol.iterator]();
  }
}

export class Ast<TType extends Component> {
  readonly #data: Array<ComponentGroup<TType>>;

  constructor(...data: Array<ComponentGroup<TType>>) {
    this.#data = data;
  }

  *iterator() {
    for (const item of this.#data) yield* item.iterator();
  }

  visited(visitor: Visitor<Component>) {
    return new Ast(...this.#data.map((g) => g.visited(visitor)));
  }
}
