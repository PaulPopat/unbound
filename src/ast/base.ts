import { Location } from "@location";

export abstract class Component {
  readonly #location: Location;

  constructor(location: Location) {
    this.#location = location;
  }

  get Location() {
    return this.#location;
  }
}

export class ComponentGroup<TComponent extends Component> {
  readonly #components: Array<TComponent>;

  constructor(...components: Array<TComponent>) {
    this.#components = components;
  }

  get Length() {
    return this.#components.length;
  }

  get First() {
    return this.#components[0];
  }

  get Last() {
    return this.#components[this.#components.length - 1];
  }

  get Location() {
    return new Location(
      this.First.Location.StartLine,
      this.First.Location.StartColumn,
      this.Last.Location.EndLine,
      this.Last.Location.EndColumn
    );
  }
}
