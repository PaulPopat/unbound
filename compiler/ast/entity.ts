import { Component, ComponentGroup } from "./base";
import { FunctionParameter, Property } from "./property";
import { Type } from "./type";
import { Statement } from "./statement";
import { Location } from "@compiler/location";
import { Expression } from "./expression";

export abstract class Entity extends Component {
  readonly #exported: boolean;

  constructor(ctx: Location, exported: boolean) {
    super(ctx);
    this.#exported = exported;
  }

  get Exported() {
    return this.#exported;
  }

  abstract get more_json(): Record<never, never>;

  get extra_json() {
    return {
      ...this.more_json,
      exported: this.#exported,
    };
  }
}

export class FunctionEntity extends Entity {
  readonly #name: string;
  readonly #parameters: ComponentGroup<FunctionParameter<Type>>;
  readonly #returns: Type | undefined;
  readonly #content: ComponentGroup<Statement>;

  constructor(
    ctx: Location,
    exported: boolean,
    name: string,
    parameters: ComponentGroup<FunctionParameter<Type>>,
    returns: Type | undefined,
    content: ComponentGroup<Statement>
  ) {
    super(ctx, exported);
    this.#name = name;
    this.#parameters = parameters;
    this.#returns = returns;
    this.#content = content;
  }

  get Name() {
    return this.#name;
  }

  get type_name() {
    return "function_entity";
  }

  get more_json() {
    return {
      name: this.#name,
      parameters: this.#parameters.json,
      returns: this.#returns?.json ?? null,
      content: this.#content.json,
    };
  }
}

export class StructEntity extends Entity {
  readonly #name: string;
  readonly #properties: ComponentGroup<Property<Type>>;

  constructor(
    ctx: Location,
    exported: boolean,
    name: string,
    properties: ComponentGroup<Property<Type>>
  ) {
    super(ctx, exported);
    this.#name = name;
    this.#properties = properties;
  }

  get type_name() {
    return "struct_entity";
  }

  get more_json() {
    return {
      name: this.#name,
      properties: this.#properties.json,
    };
  }
}

export class SchemaEntity extends Entity {
  readonly #name: string;
  readonly #properties: ComponentGroup<Property<Type>>;

  constructor(
    ctx: Location,
    exported: boolean,
    name: string,
    properties: ComponentGroup<Property<Type>>
  ) {
    super(ctx, exported);
    this.#name = name;
    this.#properties = properties;
  }

  get type_name() {
    return "schema_entity";
  }

  get more_json() {
    return {
      name: this.#name,
      properties: this.#properties.json,
    };
  }
}

export class UsingEntity extends Entity {
  readonly #name: string;

  constructor(ctx: Location, exported: boolean, name: string) {
    super(ctx, exported);
    this.#name = name;
  }

  get type_name() {
    return "using_entity";
  }

  get more_json() {
    return {
      name: this.#name,
    };
  }
}

export class ExternalFunctionDeclaration extends Component {
  readonly #name: string;
  readonly #parameters: ComponentGroup<FunctionParameter<Type>>;
  readonly #returns: Type;

  constructor(
    ctx: Location,
    name: string,
    parameters: ComponentGroup<FunctionParameter<Type>>,
    returns: Type
  ) {
    super(ctx);
    this.#name = name;
    this.#parameters = parameters;
    this.#returns = returns;
  }

  get type_name() {
    return "external_function_declaration";
  }

  get extra_json() {
    return {
      name: this.#name,
      parameters: this.#parameters.json,
      returns: this.#returns.json,
    };
  }
}

export class LibEntity extends Entity {
  readonly #name: Expression;
  readonly #content: ComponentGroup<ExternalFunctionDeclaration>;

  constructor(
    ctx: Location,
    exported: boolean,
    name: Expression,
    content: ComponentGroup<ExternalFunctionDeclaration>
  ) {
    super(ctx, exported);
    this.#name = name;
    this.#content = content;
  }

  get Name() {
    return this.#name;
  }

  get type_name() {
    return "lib_entity";
  }

  get more_json() {
    return {
      name: this.#name,
      content: this.#content.json,
    };
  }
}

export class SystemEntity extends Entity {
  readonly #content: ComponentGroup<ExternalFunctionDeclaration>;

  constructor(
    ctx: Location,
    exported: boolean,
    content: ComponentGroup<ExternalFunctionDeclaration>
  ) {
    super(ctx, exported);
    this.#content = content;
  }

  get type_name() {
    return "system_entity";
  }

  get more_json() {
    return {
      content: this.#content.json,
    };
  }
}
