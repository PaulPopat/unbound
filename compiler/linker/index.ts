import { Ast, Component } from "@compiler/ast";
import { FunctionCollectingVisitor } from "./visitor/function-collecting-visitor";
import { TypeCollectingVisitor } from "./visitor/type-collecting-visitor";
import { ReferenceExpressionVisitor } from "./visitor/reference-expression-visitor";
import { ReferenceTypeVisitor } from "./visitor/reference-type-visitor";

export function LinkUnbound(ast: Ast<Component>) {
  const function_collector = new FunctionCollectingVisitor();
  const type_collector = new TypeCollectingVisitor();

  return ast
    .visited(function_collector)
    .visited(type_collector)
    .visited(new ReferenceExpressionVisitor(function_collector.Functions))
    .visited(new ReferenceTypeVisitor(type_collector.Types));
}
