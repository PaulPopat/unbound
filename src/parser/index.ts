import { ComponentGroup, Namespace } from "@ast";
import { SplitTokens } from "./tokeniser";

export function ParseUnbound(input: string): ComponentGroup<Namespace> {
  const tokens = SplitTokens(input);
  return new ComponentGroup();
}
