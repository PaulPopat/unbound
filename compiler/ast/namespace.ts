import { Location } from "@compiler/location";
import { Component, ComponentGroup, Visitor } from "./base";
import { Entity } from "./entity";

export class Namespace extends Component {
  readonly #name: string;
  readonly #exported: boolean;
  readonly #contents: ComponentGroup<Entity>;

  constructor(
    ctx: Location,
    exported: boolean,
    name: string,
    contents: ComponentGroup<Entity>
  ) {
    super(ctx);
    this.#name = name;
    this.#exported = exported;
    this.#contents = contents;
  }

  get Name() {
    return this.#name;
  }

  get Exported() {
    return this.#exported;
  }

  get Contents() {
    return this.#contents;
  }

  get type_name() {
    return "namespace";
  }

  get extra_json() {
    return {
      name: this.#name,
      exported: this.#exported,
      contents: this.#contents.json,
    };
  }

  inner_visited(visitor: Visitor): Component {
    return new Namespace(
      this.Location,
      this.#exported,
      this.#name,
      this.#contents.type_safe_visited(Entity, visitor)
    );
  }
}
