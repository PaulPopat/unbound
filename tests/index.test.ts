import { LinkUnbound } from "@compiler/linker";
import { ParseUnbound } from "@compiler/parser";
import Fs from "fs";
import Path from "path";
import { Ast } from "../compiler/ast/base";

function try_read(path: string) {
  try {
    return Fs.readFileSync(path, "utf-8");
  } catch {
    return undefined;
  }
}

function snapshot_test(
  file_name: string,
  dirname: string,
  process: (data: string) => any
) {
  if (file_name === "snapshots") return;
  const path = Path.resolve(dirname, file_name);
  const expected_path = Path.resolve(dirname, "snapshots", file_name + ".json");

  it(file_name, () => {
    const content = Fs.readFileSync(path, "utf-8");
    const result = process(content);
    const expected = try_read(expected_path);

    if (!expected) {
      try {
        Fs.mkdirSync(Path.resolve(dirname, "snapshots"));
      } catch {}
      Fs.writeFileSync(expected_path, JSON.stringify(result, undefined, 2));
    }

    expect(result).toEqual(JSON.parse(expected ?? "{}"));
  });
}

describe("parser", () => {
  const base = Path.resolve(__dirname, "parser");
  const success = Path.resolve(base, "success");
  const error = Path.resolve(base, "errors");
  for (const file of Fs.readdirSync(success)) {
    snapshot_test(file, success, (d) => ParseUnbound(d, file).json);
  }

  for (const file of Fs.readdirSync(error)) {
    snapshot_test(file, error, (d) => {
      try {
        ParseUnbound(d, file);
      } catch (err: any) {
        return err.toString();
      }

      throw new Error("Parsing succeeded");
    });
  }
});

describe("linker", () => {
  const base = Path.resolve(__dirname, "linker");
  const success = Path.resolve(base, "success");
  const error = Path.resolve(base, "errors");
  for (const file of Fs.readdirSync(success)) {
    snapshot_test(
      file,
      success,
      (d) => LinkUnbound(new Ast(ParseUnbound(d, file))).json
    );
  }

  for (const file of Fs.readdirSync(error)) {
    snapshot_test(file, error, (d) => {
      try {
        LinkUnbound(new Ast(ParseUnbound(d, file))).json;
      } catch (err: any) {
        return err.toString();
      }

      throw new Error("Linking succeeded");
    });
  }
});
