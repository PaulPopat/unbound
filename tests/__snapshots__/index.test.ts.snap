// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser errors when namespace with no block 1`] = `
"Parsing Error:
Unexpected end of file
Line: undefined
Column: undefined"
`;

exports[`parser errors when no parameters section in a function 1`] = `
"Parsing Error:
Unexpected symbol. Expected one of '(' but received ':'
Line: 1
Column: 9"
`;

exports[`parser errors when spaces in a namespace name 1`] = `
"Parsing Error:
Unexpected symbol. Expected one of '{', '.' but received 'Name'
Line: 0
Column: 15"
`;

exports[`parser parses a empty exported schema 1`] = `
[
  {
    "contents": [
      {
        "exported": true,
        "location": {
          "last": {
            "column": 15,
            "line": 1,
          },
          "start": {
            "column": 9,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [],
        "type": "schema_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty file 1`] = `[]`;

exports[`parser parses a empty function 1`] = `
[
  {
    "contents": [
      {
        "content": [],
        "exported": false,
        "location": {
          "last": {
            "column": 4,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "parameters": [],
        "returns": {
          "location": {
            "last": {
              "column": 16,
              "line": 1,
            },
            "start": {
              "column": 13,
              "line": 1,
            },
          },
          "name": "int",
          "type": "reference_type",
        },
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty function with a single parameter 1`] = `
[
  {
    "contents": [
      {
        "content": [],
        "exported": false,
        "location": {
          "last": {
            "column": 4,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "parameters": [
          {
            "location": {
              "last": {
                "column": 15,
                "line": 1,
              },
              "start": {
                "column": 11,
                "line": 1,
              },
            },
            "name": "test",
            "type": "function_parameter",
          },
        ],
        "returns": {
          "location": {
            "last": {
              "column": 27,
              "line": 1,
            },
            "start": {
              "column": 24,
              "line": 1,
            },
          },
          "name": "int",
          "type": "reference_type",
        },
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty function with multiple parameters 1`] = `
[
  {
    "contents": [
      {
        "content": [],
        "exported": false,
        "location": {
          "last": {
            "column": 4,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "parameters": [
          {
            "location": {
              "last": {
                "column": 15,
                "line": 1,
              },
              "start": {
                "column": 11,
                "line": 1,
              },
            },
            "name": "test",
            "type": "function_parameter",
          },
          {
            "location": {
              "last": {
                "column": 28,
                "line": 1,
              },
              "start": {
                "column": 23,
                "line": 1,
              },
            },
            "name": "other",
            "type": "function_parameter",
          },
        ],
        "returns": {
          "location": {
            "last": {
              "column": 40,
              "line": 1,
            },
            "start": {
              "column": 37,
              "line": 1,
            },
          },
          "name": "int",
          "type": "reference_type",
        },
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty function with no return type 1`] = `
[
  {
    "contents": [
      {
        "content": [],
        "exported": false,
        "location": {
          "last": {
            "column": 4,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "parameters": [],
        "returns": null,
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty namespace 1`] = `
[
  {
    "contents": [],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty public struct 1`] = `
[
  {
    "contents": [
      {
        "exported": true,
        "location": {
          "last": {
            "column": 15,
            "line": 1,
          },
          "start": {
            "column": 9,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [],
        "type": "struct_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty schema 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [],
        "type": "schema_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a empty struct 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [],
        "type": "struct_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a exported empty function 1`] = `
[
  {
    "contents": [
      {
        "content": [],
        "exported": true,
        "location": {
          "last": {
            "column": 11,
            "line": 1,
          },
          "start": {
            "column": 9,
            "line": 1,
          },
        },
        "name": "Test",
        "parameters": [],
        "returns": {
          "location": {
            "last": {
              "column": 24,
              "line": 1,
            },
            "start": {
              "column": 21,
              "line": 1,
            },
          },
          "name": "int",
          "type": "reference_type",
        },
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a exported namespace 1`] = `
[
  {
    "contents": [],
    "exported": true,
    "location": {
      "last": {
        "column": 16,
        "line": 0,
      },
      "start": {
        "column": 7,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a linq 1`] = `
[
  {
    "contents": [
      {
        "content": [
          {
            "location": {
              "last": {
                "column": 10,
                "line": 2,
              },
              "start": {
                "column": 4,
                "line": 2,
              },
            },
            "type": "return_statement",
            "value": {
              "as": "item",
              "location": {
                "last": {
                  "column": 18,
                  "line": 2,
                },
                "start": {
                  "column": 11,
                  "line": 2,
                },
              },
              "over": {
                "location": {
                  "last": {
                    "column": 24,
                    "line": 2,
                  },
                  "start": {
                    "column": 20,
                    "line": 2,
                  },
                },
                "name": "self",
                "type": "reference_expression",
              },
              "type": "iterate_expression",
              "using": [
                {
                  "location": {
                    "last": {
                      "column": 12,
                      "line": 3,
                    },
                    "start": {
                      "column": 6,
                      "line": 3,
                    },
                  },
                  "type": "return_statement",
                  "value": {
                    "location": {
                      "last": {
                        "column": 27,
                        "line": 3,
                      },
                      "start": {
                        "column": 26,
                        "line": 3,
                      },
                    },
                    "parameters": [],
                    "subject": {
                      "location": {
                        "last": {
                          "column": 26,
                          "line": 3,
                        },
                        "start": {
                          "column": 18,
                          "line": 3,
                        },
                      },
                      "name": "selector",
                      "type": "reference_expression",
                    },
                    "type": "invokation_expression",
                  },
                },
              ],
            },
          },
        ],
        "exported": true,
        "location": {
          "last": {
            "column": 11,
            "line": 1,
          },
          "start": {
            "column": 9,
            "line": 1,
          },
        },
        "name": "select",
        "parameters": [
          {
            "location": {
              "last": {
                "column": 23,
                "line": 1,
              },
              "start": {
                "column": 19,
                "line": 1,
              },
            },
            "name": "self",
            "type": "function_parameter",
          },
          {
            "location": {
              "last": {
                "column": 48,
                "line": 1,
              },
              "start": {
                "column": 40,
                "line": 1,
              },
            },
            "name": "selector",
            "type": "function_parameter",
          },
        ],
        "returns": {
          "location": {
            "last": {
              "column": 78,
              "line": 1,
            },
            "start": {
              "column": 77,
              "line": 1,
            },
          },
          "type": "iterable_type",
        },
        "type": "function_entity",
      },
      {
        "content": [
          {
            "location": {
              "last": {
                "column": 10,
                "line": 8,
              },
              "start": {
                "column": 4,
                "line": 8,
              },
            },
            "type": "return_statement",
            "value": {
              "as": "item",
              "location": {
                "last": {
                  "column": 18,
                  "line": 8,
                },
                "start": {
                  "column": 11,
                  "line": 8,
                },
              },
              "over": {
                "location": {
                  "last": {
                    "column": 24,
                    "line": 8,
                  },
                  "start": {
                    "column": 20,
                    "line": 8,
                  },
                },
                "name": "self",
                "type": "reference_expression",
              },
              "type": "iterate_expression",
              "using": [
                {
                  "location": {
                    "last": {
                      "column": 12,
                      "line": 9,
                    },
                    "start": {
                      "column": 6,
                      "line": 9,
                    },
                  },
                  "type": "return_statement",
                  "value": {
                    "check": {
                      "location": {
                        "last": {
                          "column": 32,
                          "line": 9,
                        },
                        "start": {
                          "column": 31,
                          "line": 9,
                        },
                      },
                      "parameters": [],
                      "subject": {
                        "location": {
                          "last": {
                            "column": 31,
                            "line": 9,
                          },
                          "start": {
                            "column": 22,
                            "line": 9,
                          },
                        },
                        "name": "predicate",
                        "type": "reference_expression",
                      },
                      "type": "invokation_expression",
                    },
                    "else": [
                      {
                        "location": {
                          "last": {
                            "column": 14,
                            "line": 12,
                          },
                          "start": {
                            "column": 8,
                            "line": 12,
                          },
                        },
                        "type": "return_statement",
                        "value": {
                          "location": {
                            "last": {
                              "column": 19,
                              "line": 12,
                            },
                            "start": {
                              "column": 15,
                              "line": 12,
                            },
                          },
                          "name": "skip",
                          "type": "reference_expression",
                        },
                      },
                    ],
                    "if": [
                      {
                        "location": {
                          "last": {
                            "column": 14,
                            "line": 10,
                          },
                          "start": {
                            "column": 8,
                            "line": 10,
                          },
                        },
                        "type": "return_statement",
                        "value": {
                          "location": {
                            "last": {
                              "column": 19,
                              "line": 10,
                            },
                            "start": {
                              "column": 15,
                              "line": 10,
                            },
                          },
                          "name": "item",
                          "type": "reference_expression",
                        },
                      },
                    ],
                    "location": {
                      "last": {
                        "column": 15,
                        "line": 9,
                      },
                      "start": {
                        "column": 13,
                        "line": 9,
                      },
                    },
                    "type": "if_expression",
                  },
                },
              ],
            },
          },
        ],
        "exported": true,
        "location": {
          "last": {
            "column": 11,
            "line": 7,
          },
          "start": {
            "column": 9,
            "line": 7,
          },
        },
        "name": "where",
        "parameters": [
          {
            "location": {
              "last": {
                "column": 22,
                "line": 7,
              },
              "start": {
                "column": 18,
                "line": 7,
              },
            },
            "name": "self",
            "type": "function_parameter",
          },
          {
            "location": {
              "last": {
                "column": 48,
                "line": 7,
              },
              "start": {
                "column": 39,
                "line": 7,
              },
            },
            "name": "predicate",
            "type": "function_parameter",
          },
        ],
        "returns": {
          "location": {
            "last": {
              "column": 71,
              "line": 7,
            },
            "start": {
              "column": 70,
              "line": 7,
            },
          },
          "type": "iterable_type",
        },
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Linq",
    "type": "namespace",
  },
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 7,
            "line": 19,
          },
          "start": {
            "column": 2,
            "line": 19,
          },
        },
        "name": "Linq",
        "type": "using_entity",
      },
      {
        "exported": true,
        "location": {
          "last": {
            "column": 15,
            "line": 21,
          },
          "start": {
            "column": 9,
            "line": 21,
          },
        },
        "name": "User",
        "properties": [
          {
            "location": {
              "last": {
                "column": 12,
                "line": 22,
              },
              "start": {
                "column": 4,
                "line": 22,
              },
            },
            "name": "username",
            "type": "property",
          },
        ],
        "type": "struct_entity",
      },
      {
        "content": [
          {
            "location": {
              "last": {
                "column": 10,
                "line": 26,
              },
              "start": {
                "column": 4,
                "line": 26,
              },
            },
            "type": "return_statement",
            "value": {
              "location": {
                "last": {
                  "column": 23,
                  "line": 27,
                },
                "start": {
                  "column": 22,
                  "line": 27,
                },
              },
              "parameters": [
                {
                  "expression": {
                    "location": {
                      "last": {
                        "column": 40,
                        "line": 27,
                      },
                      "start": {
                        "column": 36,
                        "line": 27,
                      },
                    },
                    "struct": "User",
                    "type": "make_expression",
                    "using": [
                      {
                        "equals": {
                          "location": {
                            "last": {
                              "column": 67,
                              "line": 28,
                            },
                            "start": {
                              "column": 36,
                              "line": 28,
                            },
                          },
                          "type": "literal_expression",
                          "value": "Do I really need to say this?",
                        },
                        "location": {
                          "last": {
                            "column": 24,
                            "line": 28,
                          },
                          "start": {
                            "column": 18,
                            "line": 28,
                          },
                        },
                        "name": "username",
                        "type": "assign_statement",
                      },
                    ],
                  },
                  "location": {
                    "last": {
                      "column": 25,
                      "line": 27,
                    },
                    "start": {
                      "column": 23,
                      "line": 27,
                    },
                  },
                  "parameters": [
                    {
                      "location": {
                        "last": {
                          "column": 31,
                          "line": 27,
                        },
                        "start": {
                          "column": 27,
                          "line": 27,
                        },
                      },
                      "name": "user",
                      "type": "function_parameter",
                    },
                  ],
                  "type": "lambda_expression",
                },
              ],
              "subject": {
                "location": {
                  "last": {
                    "column": 22,
                    "line": 27,
                  },
                  "start": {
                    "column": 16,
                    "line": 27,
                  },
                },
                "name": "select",
                "type": "reference_expression",
              },
              "type": "invokation_expression",
            },
          },
        ],
        "exported": true,
        "location": {
          "last": {
            "column": 11,
            "line": 25,
          },
          "start": {
            "column": 9,
            "line": 25,
          },
        },
        "name": "sort_users",
        "parameters": [
          {
            "location": {
              "last": {
                "column": 27,
                "line": 25,
              },
              "start": {
                "column": 23,
                "line": 25,
              },
            },
            "name": "self",
            "type": "function_parameter",
          },
        ],
        "returns": null,
        "type": "function_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 18,
      },
      "start": {
        "column": 0,
        "line": 18,
      },
    },
    "name": "LinqUser",
    "type": "namespace",
  },
]
`;

exports[`parser parses a schema with a single property 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [
          {
            "location": {
              "last": {
                "column": 8,
                "line": 2,
              },
              "start": {
                "column": 4,
                "line": 2,
              },
            },
            "name": "test",
            "type": "property",
          },
        ],
        "type": "schema_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a schema with two parameters 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [
          {
            "location": {
              "last": {
                "column": 8,
                "line": 2,
              },
              "start": {
                "column": 4,
                "line": 2,
              },
            },
            "name": "test",
            "type": "property",
          },
          {
            "location": {
              "last": {
                "column": 9,
                "line": 3,
              },
              "start": {
                "column": 4,
                "line": 3,
              },
            },
            "name": "other",
            "type": "property",
          },
        ],
        "type": "schema_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a struct with a single property 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [
          {
            "location": {
              "last": {
                "column": 8,
                "line": 2,
              },
              "start": {
                "column": 4,
                "line": 2,
              },
            },
            "name": "test",
            "type": "property",
          },
        ],
        "type": "struct_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a struct with two properties 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 8,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Test",
        "properties": [
          {
            "location": {
              "last": {
                "column": 8,
                "line": 2,
              },
              "start": {
                "column": 4,
                "line": 2,
              },
            },
            "name": "test",
            "type": "property",
          },
          {
            "location": {
              "last": {
                "column": 9,
                "line": 3,
              },
              "start": {
                "column": 4,
                "line": 3,
              },
            },
            "name": "other",
            "type": "property",
          },
        ],
        "type": "struct_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;

exports[`parser parses a using entity 1`] = `
[
  {
    "contents": [
      {
        "exported": false,
        "location": {
          "last": {
            "column": 7,
            "line": 1,
          },
          "start": {
            "column": 2,
            "line": 1,
          },
        },
        "name": "Other",
        "type": "using_entity",
      },
    ],
    "exported": false,
    "location": {
      "last": {
        "column": 9,
        "line": 0,
      },
      "start": {
        "column": 0,
        "line": 0,
      },
    },
    "name": "Test",
    "type": "namespace",
  },
]
`;
